milner_desvars/wk3-notes.md

```diff
@@ -0,0 +1,294 @@
+
+So we set our new breakpoint by using `break 10`, and run the program again, once it stops in the breakpoint, we use the following command to see the value of `NODE_NAME`:
+
+`disp/s NODE_NAME`
+
+And we get the value `"navigation_goals"`, we set the format of the output by using the `/s` and obtain a string as an output, manipulating this value you can obtain the output in different formats such as int or hexadecimal.
+
+As we show this are the basic movements to start debugging a node in C++, for more commands to use in `gdb` you can see them [here](http://www.yolinux.com/TUTORIALS/GDB-Commands.html).
+
+Since we can't copy what happened in the gdb xterm, a photo that illustrates how the process looks:
+
+![gdb](media/gbd.png)
+
+### Example of debugging a python node with pdb
+
+pdb is the debugger for python that is included in the standard python library, since that library is already used in our create_autonomy repo, we don't need to install anything.
+
+To run pdb in the node we add the prefix `launch-prefix="xterm -e python -m pdb"` in the launch file to execute the pdb on an xterm.
+
+So the file would end up being like this:
+```
+<launch>
+  <node pkg="ca_tools" type="key_teleop.py" name="keyboard_teleop" output="screen" launch-prefix="xterm -e python -m pdb"/>
+</launch>
+```
+
+Launching that with roslaunch would open the xterm with pdb, and now we can begin to debug.
+
+After launching we get in the xterm the pdb, now we are going to use some commands to do the proper debugging.
+
+The first command we are going to see is `list`, using `list` loads the first lines of the code so you can have a quick scope on the file you are going to debug.
+
+In the xterm, we see these characters `->`. which are used to identify in which line we are right now.
+
+We can continue to run the program using `step` or `next`, The difference is that `step` will stop within a called function, while `next` executes called functions to only stop at the next line of the current function. 
+
+Other thing we can do is put a breakpoint, we do that using the `break` command, followed by the number of the line in which we are going to put the breakpoint:
+
+`break 14`
+
+Those are some of the basics tools to debug in pdb, you can also integrate pdb into the program code by editing the file, for more information and a table of utilities you can visit the following page:
+
+[https://www.digitalocean.com/community/tutorials/how-to-use-the-python-debugger](https://www.digitalocean.com/community/tutorials/how-to-use-the-python-debugger)
+
+As an example, this is how the xterm of `pdb` looks:
+
+![pdb](media/pdb.png)
+
+## rqt_launch
+
+rqt_launch is an rqt plugin that allows you to see and easy way to view nodes and also to run or end nodes individually.
+
+To run it, first you enter in rqt using on the console:
+
+`rqt` 
+
+After that, you select the `Launch(Experimental)` option in the Plugin tab.
+
+![rqtlaunch](media/rqtlaunch1.png)
```

Comment:
The bottom half of the image is useless so check if you can remove it. A very small part of the image shows useful information.

Reply:
Addressed